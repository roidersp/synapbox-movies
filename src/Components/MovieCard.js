import React from "react";
import { useMutation, useQuery } from "@apollo/react-hooks";

import { makeStyles, withStyles } from "@material-ui/core/styles";
import {
  Card,
  CardActions,
  CardContent,
  CardMedia,
  Typography,
  Grid,
  Fab
} from "@material-ui/core";

import { ADD_TO_CART, IS_IN_CART } from "../Actions/queries";
import { AddShoppingCart, Visibility } from "@material-ui/icons";

const useStyles = makeStyles({
  card: {
    width: "100%",
    height: "100%",
    display: "flex",
    flexDirection: "column",
    alignItems: "flex-end",
    justifyContent: "space-between"
  },
  media: {
    height: 160,
    width: "100%",
    flex: "0"
  },
  addButton: {
    fontSize: 18
  },
  seeButton: {
    marginRight: 10
  },
  title: {
    marginTop: 10
  },
  cardActions: {
    display: "flex",
    justifyContent: "space-between",
    marginBottom: 15
  }
});

const CustomIcon = withStyles(theme => ({
  root: {
    color: "#fff",
    backgroundColor: theme.palette.secondary.dark,
    padding: 8,
    "&:hover": {
      backgroundColor: theme.palette.secondary.light,
      color: "#c3c3c3"
    }
  }
}))(Fab);

const MovieCard = ({ id, title, image, price }) => {
  const classes = useStyles();

  // const {data, loading: queryLoading , error: queryError, ...lets} = useQuery(
  //   IS_IN_CART,
  //   { variables: { id: id } }
  // )

  const [mutate, { loading, error, client, called, ...rest }] = useMutation(
    ADD_TO_CART,
    {
      variables: { id: id }
      // refetchQueries: [
      //   {
      //     query: IS_IN_CART,
      //     variables: { id: id },
      //   }
      // ]
    }
  );

  const setMovie = () => {
    client.writeData({ data: { movieId: id } });
  };

  if (loading) console.log("loading mutation");

  if (error) console.log("error", error);

  if (called) {
    console.log("called");
  }

  return (
    <Grid item lg={3} md={4} sm={6} xs={12}>
      <Card className={classes.card}>
        <div style={{ width: "100%" }}>
          <CardMedia className={classes.media} image={image} title={title} />
          <CardContent>
            <Typography className={classes.title} variant="h5" component="h2">
              {title}
            </Typography>
            <Typography variant="body2" color="textSecondary" component="p">
              ${price}
            </Typography>
          </CardContent>
        </div>
        <CardActions className={classes.cardActions}>
          <Fab variant="extended" size="medium" onClick={setMovie}>
            <Visibility className={classes.seeButton} />
            Ver detalle
          </Fab>
          <CustomIcon variant="extended" size="medium" onClick={mutate}>
            {loading ? (
              <p>Loading...</p>
            ) : error ? (
              <p>Ocurrio un error</p>
            ) : (
              <AddShoppingCart className={classes.addButton} />
            )}
          </CustomIcon>
        </CardActions>
      </Card>
    </Grid>
  );
};

export default MovieCard;
