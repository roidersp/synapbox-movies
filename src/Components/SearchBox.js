import React, { useState } from "react";

import {
  Container,
  FormControl,
  InputLabel,
  OutlinedInput,
  InputAdornment,
  IconButton,
  Paper
} from "@material-ui/core";
import CloseIcon from "@material-ui/icons/Close";
import { makeStyles } from "@material-ui/core/styles";

import SearchList from "./SearchList";

const styles = makeStyles(theme => ({
  mainContainer: {
    maxWidth: 800,
    marginBottom: 20,
    position: "relative"
  },
  listContainer: {
    width: "100%",
    position: "absolute",
    top: "102%",
    left: 0,
    zIndex: 10
  },
  paperList: {
    padding: "20px 10px",
    width: "100%",
    border: `1px solid ${theme.palette.primary.main}`
  }
}));

const SearchBox = () => {
  const [searchValue, setValue] = useState("");

  const handleChange = event => {
    setValue(event.target.value);
  };

  const cleanInput = () => {
    setValue("");
  };

  const classes = styles();

  return (
    <Container className={classes.mainContainer}>
      <FormControl fullWidth variant="outlined">
        <InputLabel htmlFor="outlined-adornment-amount">Buscar</InputLabel>
        <OutlinedInput
          id="outlined-adornment-amount"
          onChange={handleChange}
          value={searchValue}
          endAdornment={
            <InputAdornment position="start">
              <IconButton onClick={cleanInput}>
                <CloseIcon />
              </IconButton>
            </InputAdornment>
          }
          labelWidth={48}
        />
      </FormControl>
      {searchValue && (
        <Container className={classes.mainContainer} fixed>
          <Paper className={classes.listContainer}>
            <SearchList searchValue={searchValue} />
          </Paper>
        </Container>
      )}
    </Container>
  );
};

export default SearchBox;
