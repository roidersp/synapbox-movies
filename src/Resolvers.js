import gql from "graphql-tag";
import { GET_CART_ITEMS } from "./Actions/queries";

export const resolvers = {
  Mutation: {
    addToCart: (_, { id }, { cache }) => {
      const { cartItems } = cache.readQuery({ query: GET_CART_ITEMS });

      const data = {
        cartItems: !cartItems.includes(id) ? [...cartItems, id] : cartItems
      };

      cache.writeQuery({ query: GET_CART_ITEMS, data });
      return data.cartItems;
    },
    deleteFromCart: (_, { id }, { cache }) => {
      const { cartItems } = cache.readQuery({ query: GET_CART_ITEMS });

      const data = {
        cartItems: cartItems.includes(id)
          ? cartItems.filter(i => i !== id)
          : cartItems
      };

      cache.writeQuery({ query: GET_CART_ITEMS, data });
      return cartItems.includes(id);
    }
  }
};

export const typeDefs = `
  extend type Mutation {
    addToCart(id: ID!)
    deleteFromCart(id: ID!)
  }
`;
