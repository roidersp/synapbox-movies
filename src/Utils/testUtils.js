import wait from "waait";
import { act } from "@testing-library/react";

export const actWait = async (amount = 0) => {
  await act(async () => {
    await wait(amount);
  });
};
