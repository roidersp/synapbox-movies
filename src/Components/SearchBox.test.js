import React from "react";
import { MockedProvider } from "@apollo/react-testing";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { actWait } from "../Utils/testUtils";
import {
  GET_SEARCH_MOVIES,
  GET_CART_ITEMS,
  GET_MOVIE_ID
} from "../Actions/queries";
import MovieCard from "./MovieCard";
import { InMemoryCache } from "apollo-cache-inmemory";
import { resolvers } from "../Resolvers";
import { renderHook, act } from "@testing-library/react-hooks";

import SearchBox from "./SearchBox";

const mocks = [
  {
    request: { query: GET_SEARCH_MOVIES, variables: { searchValue: "the" } },
    result: {
      data: {
        movies: {
          aggregate: {
            count: 0
          },
          edges: [],
          pageInfo: {
            hasNextPage: true,
            hasPreviousPage: false,
            startCursor: "ck49086bqsnup0b09d14atpb2",
            endCursor: "ck490n16isok60b090jqnbv20"
          }
        }
      }
    }
  }
];

describe("<SearchBox />", () => {
  it("snapshot SearchBox", () => {
    const Wrapper = render(
      <MockedProvider
        mocks={mocks}
        addTypename={false}
        resolvers={{}}
        cache={{}}
      >
        <SearchBox />
      </MockedProvider>
    );
    expect(Wrapper).toMatchSnapshot();
  });

  it("Input onChange", async () => {
    const { getByTestId, getByLabelText } = render(
      <MockedProvider mocks={mocks} addTypename={false} resolvers={{}}>
        <SearchBox />
      </MockedProvider>
    );

    const Input = getByLabelText("Buscar");

    fireEvent.change(Input, { target: { value: "the" } });
    await actWait();

    expect(getByTestId("SearchContainer")).toBeTruthy();
  });

  it("Clean input", async () => {
    const { getByTestId, getByLabelText } = render(
      <MockedProvider mocks={mocks} addTypename={false} resolvers={{}}>
        <SearchBox />
      </MockedProvider>
    );

    fireEvent.click(getByTestId("cleanInput"));
    await actWait();

    expect(getByLabelText("Buscar").value).toBe("");
  });
});
