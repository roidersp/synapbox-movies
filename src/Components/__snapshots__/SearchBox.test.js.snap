// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox /> snapshot SearchBox 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root makeStyles-mainContainer-1 MuiContainer-maxWidthLg"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
            for="outlined-adornment-amount"
          >
            Buscar
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
            data-testid="Input"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              id="outlined-adornment-amount"
              type="text"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="cleanInput"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-99 MuiOutlinedInput-notchedOutline"
              style="padding-left: 30px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-100"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root makeStyles-mainContainer-1 MuiContainer-maxWidthLg"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="outlined-adornment-amount"
        >
          Buscar
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
          data-testid="Input"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
            id="outlined-adornment-amount"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              data-testid="cleanInput"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-99 MuiOutlinedInput-notchedOutline"
            style="padding-left: 30px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-100"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
